### DISCUSSION TITLE
SQLite + journal_mode=wal : prisma connection management and WAL file

### DISCUSSION THREAD

**Discussion by 0gust1** [prisma/prisma#15966] [Q&A]
Hi,

We use SQLite with prisma in the context of a (sveltekit-powered) webApp.

In order to make it work correctly on our cloud environment (Azure), we had to activate the WAL (write ahead logging) journal mode on SQLite. It seems it can also increase the SQLite performance on some contexts.

Refs: 
- https://sqlite.org/wal.html
- https://www.sqlite.org/tempfiles.html#write_ahead_log_wal_files
- https://github.com/prisma/prisma/issues/3303

In our app, we have the following files `application.sqlite` and `application.sqlite-wal`

According to the references, this WAL file is reconciliated with the main sqlite file when the SQLite connection is closed properly. But, in our context, I don't see this: the WAL file is still growing and not reconciliated with the main DB file, even across app stop/start cycles.

I don't like this (maybe naively), in the perspective of backups, having a portable db, maintenance ease, etc.

**Our app seems to fit the [long running application case](https://www.prisma.io/docs/guides/performance-and-optimization/connection-management#prismaclient-in-long-running-applications).**

Prisma documentation clearly states that we should avoid calling explicitly the `$disconnect()` method of PrismaClient (so we did, and followed the pattern showed in the doc).

### Question:

**Should we call `$disconnect()` somewhere in our app to merge the wal file and the main db file ? If so, where and how ?** 

### Further references:
- https://www.theunterminatedstring.com/sqlite-vacuuming/
- https://github.com/groue/GRDB.swift/issues/739 (user @mallman seems to have the same concerns as me)

**Comment by nurul3101** [comment]
Hello @0gust1 👋 

Thank you for providing references and explaining your use case thoroughly.

Are you facing any performance issues due to the `sqlite-wal` file?
I would recommend trying out calling `$disconnect` explicitly to observe the behaviour and check if the wal file gets removed or not.
Disconnect would close all the database connections and as the [reference](https://www.sqlite.org/tempfiles.html#write_ahead_log_wal_files) mentions:

> The WAL file is created when the first connection to the database is opened and is normally removed when the last connection to the database closes. However, if the last connection does not shutdown cleanly, the WAL file will remain in the filesystem and will be automatically cleaned up the next time the database is opened.

that the wal file should be removed once all the database connections are closed cleanly, I would expect that behaviour.
Let me know if that doesn't happen, in that case we can open a [Bug Report](https://github.com/prisma/prisma/issues/new?assignees=&labels=kind%2Fbug&template=bug_report.yml) so that our engineering team could investigate this.

I am not exactly sure if Sveltekit is similar to Next.js API routes or not, but for invoking the `$disconnect()` I would say you can do it after the API sends back the response to the client? The subsequent request would automatically reconnect to the database.

**Comment by 0gust1** [comment]
> Are you facing any performance issues due to the sqlite-wal file?

not at the moment. It was more a "let's keep things clean" thought. Also, a wal file is also bigger (due to its log nature) than its reconciliated counterpart. The filesize won't be so much a problem on the DB I'm currently working on (applicative DB), but we plan also to use SQLite for complex scientific datasets.

> I am not exactly sure if Sveltekit is similar to Next.js API routes or not

Yes, you can consider that Sveltekit and Next.js work the same at runtime (and now, their top maintainers work for the same company – Vercel :-D )

Thanks for your advice, I'll try and keep you posted!  
I plan two things: 
- do as you suggested (but with wariness, as in our case, the sqlite file is on a CIFS/SMB network drive)
- try to find a way to make a clean disconnect in code that is ran once (like just after the migrations, and before app start)

**Reply by anitacung** [reply]
👋 Hi @0gust1, 

I'm going through pending GitHub Discussions to check if you need any help. I also noticed that you opened another discussion about “Unable to run prisma migrations on Azure app service” in this post here: https://github.com/prisma/prisma/discussions/15859 and it has been resolved. 

Is this current discussion also resolved? If there's anything we can help you with, we'd love to get you in touch with our Product team.

**Reply by 0gust1** [reply]
> Is this current discussion also resolved

Not at the moment. I'll keep the team updated on my findings.

**Reply by anitacung** [reply]
Hey @0gust1, 

Thanks for following up! If you ever have any questions or issues, feel free to reach out. In the meantime, you can join our community on Slack where you can network with other Prisma users or our team. https://slack.prisma.io/

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
